fix: Initial support for open-next v2 and ISR

kevin-mitchell opened the pull request:
See: https://github.com/jetbridge/cdk-nextjs/issues/119

Updates the server handler to include cache bucket configuration, **which is likely not correct as I'm just using the static asset bucket temporarily**. I need to look at how SST does this and what the spirit of this should be, e.g. if a separate bucket would be more appropriate, a folder, etc.

This also adds the ISR handling logic. I haven't actually tested this yet (am not 100% sure how to).

**This is a WIP**.

Fixes #119 

## TODO

1. ~~Confirm if we should make a new bucket for cache or use an existing bucket~~
2. ~~Confirm permissions appropriate for this bucket between nextjs server and the bucket~~
3. Confirm if ISR feature is working and what else may be needed
4. ~~Make sure I didn't break anything along the way~~
5. I'd like to add descriptions to other functions

bestickley commented:
@kevin-mitchell, thanks for starting on this! One way to test this would be to deploy the Next.js 13 App Dir Example from Vercel [here](https://github.com/vercel/app-playground) and ensure everything works as expected. I think that example showcases many features we'd want to test against. Not sure about pages directory as I don't have as much experience with those features.

bestickley commented on src/NextjsRevaluation.ts:
Should this be `NextjsRevalidation` instead? That would align with open-next's architecture diagram [here](https://github.com/serverless-stack/open-next/blob/main/README.md#recommended-infrastructure-on-aws).

kevin-mitchell replied:
Yes for sure 😂 - I was tired when I did this and the word looked weird to me but I couldn't be bothered in the moment to figure out why. I'll fix this.

bestickley commented on src/NextjsRevaluation.ts:
```diff
@@ -0,0 +1,63 @@
+      code,
+      runtime: Runtime.NODEJS_18_X,
+      timeout: Duration.seconds(30),
+      // This, I think, is supposed to be the VPC or VPC Subnet config (https://github.com/serverless-stack/sst/blob/master/packages/sst/src/constructs/NextjsSite.ts#L59C5-L59C17)
```
I don't see why this function would need a VPC. Although, we maybe should add props to enable a developer to add this function into a VPC if they want to but I'm not sure why they would.

kevin-mitchell replied:
I'm not actually sure either to be honest. Honestly I don't know, can an SQS queue be "in" a VPC? I'd guess that's not relevant. I'll remove this for now.

bestickley replied:
@kevin-mitchell, SQS doesn't run in a VPC. It's a managed service so you don't have control over where it runs.

kevin-mitchell replied:
👍 , that was my understanding

bestickley commented on src/NextjsRevaluation.ts:
```diff
@@ -0,0 +1,63 @@
+      fifo: true,
+      receiveMessageWaitTime: Duration.seconds(20),
+    });
+    const consumer = new Function(this, 'RevalidationFunction', {
```
This function will also need permissions to update cache s3 bucket.

kevin-mitchell replied:
I'm not so sure about that actually (obviously please correct me though!) - [from what I can tell looking at open-next](https://github.com/serverless-stack/open-next/blob/274d446ed7e940cfbe7ce05a21108f4c854ee37a/packages/open-next/src/adapters/revalidate.ts#L18) it looks like the revalidation function just makes an HTTP request to the page. I don't know if I personally love this (vs the alternative of actually doing the work to generate the content required directly), but assuming this is indeed the case the revalidation function just needs access to the internet I think?

kevin-mitchell replied:
Oh, worth mentioning that in my view open-next's architecture diagram isn't really clear about this, or is just flat out wrong? There should ben arrow pointing back to the server function or something maybe?

bestickley replied:
@kevin-mitchell, I was wrong. Revalidation function should send HEAD requests to Static Assets bucket, not cache bucket so that Next.js re-creates them. open-next docs:
> Create a Lambda function using the code in the .open-next/revalidation-function folder, with the handler index.mjs.

Also, create an SQS FIFO queue, and set it as the event source for this function.

This function polls the queue for revalidation messages. Upon receiving a message, the function sends a HEAD request to the specified route for its revalidation.

However, a cache bucket does need to be created as noted in open-next docs:
> Create an S3 bucket and upload the content in the .open-next/cache folder to the root of the bucket. If you need to upload the files to a subfolder within the bucket, [refer to this section](https://github.com/serverless-stack/open-next/blob/main/README.md#reusing-same-bucket-for-asset-and-cache).

There are two types of caches in the .open-next/cache folder:

Route cache: This cache includes html and json files that are prerendered during the build. They are used to seed the revalidation cache.
Fetch cache: This cache includes fetch call responses, which might contain sensitive information. Make sure these files are not publicly accessible

kevin-mitchell replied:
There is new a separate cache bucket.

bestickley commented on src/Nextjs.ts:
```diff
@@ -134,6 +140,7 @@ export class Nextjs extends Construct {
       tempBuildDir,
       nextBuild: this.nextBuild,
       lambda: props.defaults?.lambda,
+      cacheBucket: this.staticAssetBucket,
```
we should create a separate bucket here, maybe `CacheBucket`. See diagram [here](https://github.com/serverless-stack/open-next/blob/main/README.md#recommended-infrastructure-on-aws) how there are 2 different buckets.

kevin-mitchell replied:
I'll create a new bucket. I'm slightly struggling with how much of `NextjsAssetsDeployment` to refactor so I can reuse it for this `CacheBucket`. 

> ...Create an S3 bucket and **upload the content in the .open-next/cache folder to the root of the bucket**...

I'm leaning towards slightly refactoring NextjsAssetsDeployment so that it can be re-used, because I'm assuming (given [this open-next documentation says that it's OK to share buckets](https://github.com/serverless-stack/open-next/blob/main/README.md#reusing-same-bucket-for-asset-and-cache)) all other settings could be shared (e.g. cache configuration). I actually don't have a great sense here, perhaps adding caching configuration to this bucket is confusing because it's not needed. 

Obviously open to direction / advice here!

kevin-mitchell replied:
Added a separate bucket. Deployment is much simpler than the static asset bucket (there is a "rewriter" for that, I'm not sure if that's something we need for the cache bucket as well).

kevin-mitchell replied:
I don't believe so, at least the docs don't mention it.

bestickley commented on src/NextjsLambda.ts:
```diff
@@ -90,6 +102,9 @@ export class NextJsLambda extends Construct {
     this.lambdaFunction = fn;
 
+    // todo: once we figure out the correct S3 bucket, make sure permissions are appropriate.
+    props.cacheBucket.grantReadWrite(fn);
```
In the spirit of least privilege, the server function should only need to read from cache, right? So this should be `grantRead`

kevin-mitchell replied:
I'm not sure. I have to look more about what is actually writing to the cache. I'm slightly confused because I don't see `S3Cache` referenced anywhere, which in theory is what is writing to the cache. I'm not sure where the hook is here, who is actually writing to the cache. But I suspect it's the web server. I can figure this out I just am not sure at the moment.

khuezy replied:
https://github.com/serverless-stack/open-next/blob/main/packages/open-next/src/adapters/server-adapter.ts#L204

We give nextjs a custom cache handler to get/set (read/write), so I believe we need both.

kevin-mitchell replied:
OK, so I'll leave to read/write. One thing I'm not certainly about is if there is a reason this project is applying specific policies, e.g. https://github.com/jetbridge/cdk-nextjs/blob/main/src/NextjsS3EnvRewriter.ts#L80C1-L93C9 vs the `props.cacheBucket.grantReadWrite(fn);` syntax I used? Is there a preference for me to use the more verbose syntax?

bestickley replied:
@kevin-mitchell, I think using `grantReadWrite` is ok.

khuezy commented on src/ImageOptimizationLambda.ts:
```diff
@@ -43,15 +43,16 @@ export class ImageOptimizationLambda extends Function {
       handler: 'index.handler',
       runtime: LAMBDA_RUNTIME,
       architecture: Architecture.ARM_64,
+      description: 'Next.js Image Optimization Function',
```
https://dev.to/dashbird/4-tips-for-aws-lambda-optimization-for-production-3if1

Can you change the default memory size to 1536mb please? That is faster and cheaper than 1024

kevin-mitchell replied:
Does this apply to other functions as well? e.g. https://github.com/kevin-mitchell/cdk-nextjs/blob/e60c3bfb5dc00d1871cea18de6ad3b91b29c1c79/src/NextjsS3EnvRewriter.ts#L76 or https://github.com/kevin-mitchell/cdk-nextjs/blob/e60c3bfb5dc00d1871cea18de6ad3b91b29c1c79/src/NextjsLambda.ts#L90 ?

I would think it would be appropriate to make a constant for this value and share it across Lambda functions. Based on the linked article I'm assuming it would be best to have them all at 1536.

khuezy replied:
Yes that'll be a good idea.

khuezy commented on src/Nextjs.ts:
```diff
@@ -113,16 +121,24 @@ export class Nextjs extends Construct {
     this.staticAssetBucket =
       props.defaults?.assetDeployment?.bucket ??
-      new s3.Bucket(this, 'Bucket', {
+      new s3.Bucket(this, 'AssetBucket', {
```
Is it redundant to have "Bucket"? "Assets" is better?

khuezy commented on src/Nextjs.ts:
```diff
@@ -113,16 +121,24 @@ export class Nextjs extends Construct {
+
+    // create cache bucket
+    this.cacheBucket =
+      props.defaults?.assetDeployment?.bucket ??
```
using "assetDeployment.bucket" might be misleading if we're separating out the buckets for assets and cache.

kevin-mitchell replied:
Certainly agree. One thing I'm not sure about is whether we need additional deployment options. In https://github.com/kevin-mitchell/cdk-nextjs/blob/e60c3bfb5dc00d1871cea18de6ad3b91b29c1c79/src/Nextjs.ts#L32  there are all sorts of options about the deployment, but to be honest I'm not 100% sure how this works when re-deploying over something already in production. `cache` should be temporal I'd think, and safe to be destroyed, but I also know large production sites with tends of thousands of pages it can be extremely expensive to just throw away all cache. 

For now, for this PR I'll take the approach of "KISS" and just make a simple cache bucket unless somebody tells me otherwise. This would be something to improve on in another PR potentially. Obviously open to a better option though!

khuezy commented on src/constants.ts:
```diff
@@ -5,3 +5,5 @@ export const DEFAULT_STATIC_MAX_AGE = Duration.days(30).toSeconds();
 
 export const LAMBDA_RUNTIME = Runtime.NODEJS_18_X;
+
+export const DEFAULT_LAMBA_MEMORY = 1536;
```
Linking to the article would be useful for future reference of this magic number

khuezy replied:
tl;dr: 1536mb costs 1.5x but runs twice as fast for most scenarios.

kevin-mitchell commented:
Current version of this branch is _mainly_ working, however using the Next.js demo when I attempt to navigate through the client context page I'm getting an error about a missing key in the cache.

clicking "electronics" on this page
![Screenshot 2023-07-29 at 1 10 20 PM](https://github.com/jetbridge/cdk-nextjs/assets/21270844/02527705-338a-4ec4-8d11-e3c993989c45)

results in
![Screenshot 2023-07-29 at 1 10 28 PM](https://github.com/jetbridge/cdk-nextjs/assets/21270844/2f1397f5-a31b-4f33-810e-9ef8ad2ef7c7)

And I _think_ this is the corresponding error. I have to look to see what is wrong here.

```
2023-07-29T17:01:21.900Z	e58948db-e51e-4df8-926a-616be4780b38	ERROR	{
  clientName: 'S3Client',
  commandName: 'GetObjectCommand',
  input: {
    Bucket: 'cdk2stack-webcache4214db24-1eiepcnix6r5t',
    Key: '__fetch/zu8Tv8Vb6COL0vbxjUL0N/99afeb833f222f564e02a181c4c851b948d3afac5d6ade910de49fefee39bfb2'
  },
  error: NoSuchKey: The specified key does not exist.
      at th (/var/task/cache.cjs:31:179555)
      at Nae (/var/task/cache.cjs:31:147881)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async /var/task/cache.cjs:12:44385
      at async /var/task/cache.cjs:12:4923
      at async /var/task/cache.cjs:13:25781
      at async /var/task/cache.cjs:33:49147
      at async /var/task/cache.cjs:1:20750
      at async Bh.getFetchCache (/var/task/cache.cjs:33:192559)
      at async IncrementalCache.get (/var/task/node_modules/.pnpm/next@13.4.8_@babel+core@7.22.1_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/server/lib/incremental-cache/index.js:218:27) {
    '$fault': 'client',
    '$metadata': {
      httpStatusCode: 404,
      requestId: 'MSPGT59JPT2K7M34',
      extendedRequestId: 'x6j4PSr3NhAhKw/8LFEjc+BnjDSrafVPTRxUJPShzbz8rRbiJcsKsfyXNEByCelGGH5HW6z+ZTs=',
      cfId: undefined,
      attempts: 1,
      totalRetryDelay: 0
    },
    Code: 'NoSuchKey',
    Key: '__fetch/zu8Tv8Vb6COL0vbxjUL0N/99afeb833f222f564e02a181c4c851b948d3afac5d6ade910de49fefee39bfb2',
    RequestId: 'MSPGT59JPT2K7M34',
    HostId: 'x6j4PSr3NhAhKw/8LFEjc+BnjDSrafVPTRxUJPShzbz8rRbiJcsKsfyXNEByCelGGH5HW6z+ZTs='
  },
  metadata: {
    httpStatusCode: 404,
    requestId: 'MSPGT59JPT2K7M34',
    extendedRequestId: 'x6j4PSr3NhAhKw/8LFEjc+BnjDSrafVPTRxUJPShzbz8rRbiJcsKsfyXNEByCelGGH5HW6z+ZTs=',
    cfId: undefined,
    attempts: 1,
    totalRetryDelay: 0
  }
}
```

kevin-mitchell commented:
I switched to using the explicit examples called out here: https://github.com/serverless-stack/open-next/blob/main/README.md#example

I could use some input on the CloudFront Distribution setup. What I have "works" and seems to have the same behavior as SST (without digging too deep - IMHO a 404 should be cached, but both the current branch here [as well as SST returns a "Error from cloudfront" x-cache header](https://d1gwt3w78t4dm3.cloudfront.net/page-does-not-exist)), but I should look closer to see what the "right" thing to do is.

That said, SOMETHING isn't tuned well because the SSG pages are serving very old Cloudfront caches. I'm not sure if this issue predates this PR because I never actually have used `cdk-nextjs` when it was working (before the open-next v2 update).

I'm copy / pasting some notes I took, although the 404 is fixed as noted above by removing the origin group and just replacing it will a direct call to the web server.


# Issues
I have never (successfully) been able to use this CDK construct / project, so I’m not sure which of these are new, which are related to the changes in this PR, and 

## Error related to SST in logs for Server Function

```
2023-07-30T21:15:15.332Z	e6fba54c-e62a-4d70-9b9f-52758eee431c	ERROR	Error: Cannot access bound resources. This usually happens if you are using an older version of SST. Please update SST to the latest version.
    at Object.get (file:///var/task/node_modules/.pnpm/sst@2.16.3/node_modules/sst/node/util/index.js:27:27)
    at /var/task/.next/server/pages/api/auth/[...nextauth].js:51:75
```

Not sure why this is happening, but it seems “random” (most requests don’t see it). 


## Cache isn’t behaving ideally for SSG
1. Do a new deployment of site
2. SSG page is still cached, so shows the wrong timestamp
3. Adding query parameters to “bust” cache shows expected SSG timestamp

Maybe this is working as intended, but it seems like the TTL is either too long or a cache invalidation should happen as part of the deployment, OR some documentation somewhere should say “you have to invalidate the cache yourself on deployment.”

Or I’m just missing something (very possible)

## 404 pages seem to serve 403 XML response from S3

**NOTE THIS WAS AT LEAST PARTIALLY FIXED** - I removed the default origin group with the S3 fallback with a direct call to the Lambda server which is I believe what SST also does.

I’m not 100% this is what’s actually happening, maybe it’s not S3. But it at least seems involved. One header is “Server: AmazonS3” 

I’m not sure how this is intended to work, or if this is coming from the cache bucket because the cache bucket is misconfigured, etc.

```
<Error>
<script/>
<Code>AccessDenied</Code>
<Message>Access Denied</Message>
<RequestId>63DM002111YMJTDN</RequestId>
<HostId>bQZYzcGFF5wf6CCiU2VzdkUgD0q535TC6lAB9fwmLjp/MRYGmmbx4hU2gxyhPDePqBgiye7UZhM=</HostId>
</Error>
```

**Note that the server lambda is triggered, and seems to generate the correct response** - so it’s something about either the Cloud Front Distribution or something else about the setup that’s causing the wrong result to be sent.

In other words, the result of`debug("ServerResponse data", { statusCode, headers, isBase64Encoded, body });`  here https://github.com/serverless-stack/open-next/blob/274d446ed7e940cfbe7ce05a21108f4c854ee37a/packages/open-next/src/adapters/server-adapter.ts#L131 is

```
2023-07-30T22:43:08.050Z	f1f2aecb-9b57-4598-a4cd-9e40cb1cfee0	INFO	ServerResponse data {
  statusCode: 404,
  headers: [Object: null prototype] {
    'x-powered-by': 'Next.js',
    etag: '"xl782hlhkj1o3"',
    'content-type': 'text/html; charset=utf-8',
    'content-length': 2165
  },
  isBase64Encoded: false,
  body: '<!DOCTYPE html><html><head><meta charSet="utf-8"/><meta name="viewport" content="width=device-width"/><link rel="icon" href="/favicon.ico"/><meta name="description" content="Learn how to build a personal website using Next.js"/><meta property="og:image" content="https://og-image.vercel.app/Next.js%20Sample%20Website.png?theme=light&amp;md=0&amp;fontSize=75px&amp;images=https%3A%2F%2Fassets.vercel.com%2Fimage%2Fupload%2Ffront%2Fassets%2Fdesign%2Fnextjs-black-logo.svg"/><meta name="og:title" content="Next.js Sample Website"/><meta name="twitter:card" content="summary_large_image"/><meta name="next-head-count" content="7"/><link rel="preload" href="/_next/static/css/52270e7d977c310f.css" as="style"/><link rel="stylesheet" href="/_next/static/css/52270e7d977c310f.css" data-n-g=""/><link rel="preload" href="/_next/static/css/34339a76372c2b69.css" as="style"/><link rel="stylesheet" href="/_next/static/css/34339a76372c2b69.css" data-n-p=""/><noscript data-n-css=""></noscript><script defer="" nomodule="" src="/_next/static/chunks/polyfills-78c92fac7aa8fdd8.js"></script><script src="/_next/static/chunks/webpack-8fa1640cc84ba8fe.js" defer=""></script><script src="/_next/static/chunks/framework-6698976aa0ea586d.js" defer=""></script><script src="/_next/static/chunks/main-e3ffb9c21b7073be.js" defer=""></script><script src="/_next/static/chunks/pages/_app-b0f7a974fa14c156.js" defer=""></script><script src="/_next/static/chunks/524-6309da801f10dcdd.js" defer=""></script><script src="/_next/static/chunks/pages/404-eeaaba92ee717891.js" defer=""></script><script src="/_next/static/Pe7Ceh4Ds1YSzL2f7c72U/_buildManifest.js" defer=""></script><script src="/_next/static/Pe7Ceh4Ds1YSzL2f7c72U/_ssgManifest.js" defer=""></script></head><body><div id="__next"><div class="layout_container__fbLkO"><main><article><h1>404</h1></article></main><div class="layout_backToHome__9sjx_"><a href="/">← Back to home</a></div></div></div><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{}},"page":"/404","query":{},"buildId":"Pe7Ceh4Ds1YSzL2f7c72U","nextExport":true,"autoExport":true,"isFallback":false,"scriptLoader":[]}</script></body></html>'
}
```

bestickley commented on src/NextjsDistribution.ts:
```diff
@@ -461,15 +450,18 @@ export class NextjsDistribution extends Construct {
       certificate: this.certificate,
       defaultBehavior: {
-        origin: fallbackOriginGroup, // try S3 first, then lambda
+        origin: new origins.HttpOrigin(Fn.parseDomainName(fnUrl.url), {
```
@kevin-mitchell, why are you changing the default origin behavior? This seems to be linked to the S3 XML error message you're seeing that you described.

kevin-mitchell replied:
@bestickley I changed it _to_ fix the issue with the 404/403s. Basically I was getting 403 errors that looked like S3 issues _before_ I changed this. Then I was trying to figure out what was going on, noted that what SST has for their defaultBehavior origin was quite different from this project, I updated to roughly match the simpler config they had and the issue went away.

THAT SAID, I am absolutely _not_ advocating for this change, I'm sure there is a great reason the origin was setup with the origin group like it was originally with the fallback origin group with the S3 fallback. I honestly just wasn't and am not sure what it is supposed to do. 

It seems the fallback default origin setup was (maybe?) for static resources, so that if the Next.js app returned a 404 that the S3 bucket would be checked for any static assets associated with it. Maybe that logic still makes sense, or maybe because of the new caching "system" in place it is adequate because the server already is checking the cache bucket for a cached resource.

bestickley replied:
@kevin-mitchell, got it. Thanks for explaining. This change will likely break support for Lambda Function URL IAM Auth since the URL needs to be signed, so I'd prefer to keep it like it was. It seems to me like using an `HttpOrigin` of the function URL or making the primary origin the function origin should result in the same thing.

Do you know why the previous configuration wasn't working?

bestickley replied:
@kevin-mitchell, I think that comment is wrong. See the original comment above the declaration of `fallbackOriginGroup`:

> // default handler for requests that don't match any other path:
    //   - try lambda handler first (/some-page, etc...)
    //   - if 403, fall back to S3
    //   - if 404, fall back to lambda handler
    //   - if 503, fall back to lambda handler

Since SST doesn't use fallback behavior to get assets from S3, I think we can assume that the server handler will handle fetching assets from S3 (but not 100% sure). So how about setting `defaultBehavior.origin` to `serverFunctionOrigin` to handle all traffic?

bestickley replied:
@kevin-mitchell, I apologize. You're correct. I didn't see the definition of `serverFunctionOrigin` when reviewing. We should be all good to go.

kevin-mitchell replied:
So after all of this, I reverted this change (see: https://github.com/jetbridge/cdk-nextjs/pull/125/commits/cec5e3232d8976746b7575c951f405a4804aa731)

I didn't realize before, but SST is generating the additional behaviors to cover the `/public` files at runtime. That's why they don't have the fallback. Rather than add confusion / noise, I'll leave the existing behavior for now untouched.

revmischa replied:
Personally I do not like the SST approach to /public file handling because it creates a behavior for each file. It doesn't scale. I think it means if you have over 25 files in the root of /public your app cannot be deployed. Also it has bugs like if your file has a character like `(` it won't deploy either.

khuezy replied:
It's definitely a trade off at build time vs runtime. You have to eat the 💩 🥪 either way.

kevin-mitchell replied:
I also prefer this more elegant "if you're using AWS, use AWS" approach. I think. :).

That said I am still uncertain as to why the 404 is served as a 403 from CloudFront.

I _think_ I deployed a clean / fresh version of this project's `main` branch, along with `open-next` v 1.4, and I'm fairly certain I got the same 404 behavior. Any chance somebody would be willing or able to sanity check me? Is there something "special" about this test https://github.com/serverless-stack/open-next/tree/main/example (this is what I'm testing with currently).

khuezy replied:
But remember, there's also a runtime penalty: 1) run nextjs server, page doesn't exist 2) look at S3 doesn't exist 3) fallback to 404. It might be negligible though.

kevin-mitchell replied:
Question: Is there a world in which either:

1. I add a third fallback behavior, so the behavior is

Hit ServerHandler, if 403/404/503
Fallback to S3, if 403/404/503
Fall back to ServerHandler again

The original comment @bestickley called out makes it seem like this is how it is SUPPOSED to work already:

```
// default handler for requests that don't match any other path:
// - try lambda handler first (/some-page, etc...)
// - if 403, fall back to S3
// - if 404, fall back to lambda handler
// - if 503, fall back to lambda handler
```
but at least for me it seems like the fallback of S3 just fails and that's the end of it

_or_

2. Make the S3 bucket the first origin, then fallback to ServerHandler?

The second option is a bit expensive because it adds overhead of checking S3 to every single request.

revmischa replied:
I designed CDK-NextJS to first hit S3, then if nothing exists, fall back to lambda.
However, this fallback handler is only used for routes that can't quickly be identified and routed to the correct place. For example if the route is under /api, /_next/image, etc then we don't need the fallback behavior.  So API and static files should be unaffected. If you have more information about your app, you can add your own `additionalBehaviors` as well to bypass the fallback origin system. You can also tell cloudfront to cache responses (perhaps with the new stale-while-revalidate) so it doesn't have to go to the origin every time.

I haven't measured the latency but we're talking about shaving off, I would guess, very low double-digit milliseconds in only a subset of requests.

khuezy replied:
iirc, we did do the S3 first but reversed it due to the latency.

revmischa replied:
> iirc, we did do the S3 first but reversed it due to the latency.

Yeah I thought I remembered doing S3 first, then went to check the code, and saw that is not presently the case...

Does the next handler return 403? Also doesn't the latest version of SST use a 502 or something instead? Any idea why?

khuezy replied:
I honestly don't remember what caused the problem, but I recall getting about ~300ms in extra latency because of the fallback stuff.

open-next return 503 to fail immediately I think: https://github.com/serverless-stack/open-next/blob/main/packages/open-next/src/adapters/server-adapter.ts#L317

Per Frank: https://discord.com/channels/983865673656705025/1027265626085019769/1095464697978957875

kevin-mitchell replied:
https://github.com/jetbridge/cdk-nextjs/blame/d3cee987d24674b0ccbc43a713ad665ddef22b15/src/NextjsDistribution.ts#L336-L344

vs

khuezy replied:
Can you try w/ S3 first then server to see if you can reproduce the latency issue?

kevin-mitchell replied:
I can look. I suspect it's a real thing. Even if it's a tiny bit of overhead, it's on every request. I also want to try out the app-sandbox to make sure I see the same behavior.

For larger / production / "enterprise" situations I would think people tune this behavior to optimize for what's most important to them.

khuezy replied:
lol, very impressive chatGPT scifi! Off tangent, if you like hard sci fi and would like to alpha read my novel, hit me up!

revmischa replied:
> I can look. I suspect it's a real thing. Even if it's a tiny bit of overhead, it's on every request. I also want to try out the app-sandbox to make sure I see the same behavior.

It's not every request; the static files and API paths don't have fallback origins.

Another option is to create an edge function that does something smart, like looking at the manifests NextJS spits out to determine where to route the request. Should be possible and likely similar to what Vercel does.

kevin-mitchell replied:
> It's not every request; the static files and API paths don't have fallback origins.

Right - I should have said "every request that isn't...

```
 'api/*': lambdaBehavior,
  '_next/data/*': lambdaBehavior,
  // dynamic images go to lambda
  '_next/image*': imageBehavior,
  // known static routes
  // it would be nice to create routes for all the static files we know of
  // but we run into the limit of CacheBehaviors per distribution
  '_next/*': staticBehavior,
```
"

I wrote a VERY simple [locust](https://locust.io/) script to do some basic testing. **In the end, I'm not sure my testing methodology was enough to give me a ton of confidence that the test isn't flawed** - either that, or the test doesn't show a TON.

I did two things:

1. I used the Next.js app-playground project and generated 500 SSG pages, then I ran a load test against these pages (both with a fresh / empty cache as well as a pre-warmed cache) with both the S3 bucket as primary as well as the Lambda as primary
2. I ran the same tests, but against non SSG pages

Here is the locust test:

```
// locust.py
import time
import json
import random
from faker import Faker
fake = Faker()

from locust import HttpUser, task, between

class WebsiteUser(HttpUser):
  wait_time = between(1,5)

  def on_start(self):

    print ("done")
    return super().on_start()

  @task
  def list_page(self):
    url = "/ssg/" + str(random.randint(1, 500))    
    result = self.client.get(
            url=url,
            name="Get SSG Page",
            headers={
                "Accept": "text/html",
            },
            
        )
    print(result)
    ```
Ran with (for example) `locust -f locust/locust.py -u 15 -r 5 --run-time 300s --headless --host https://d2sstn6osjz1uu.cloudfront.net --html 300_second_GENERATED_pa
ges_lambda_first.html`    
 
Here are the test results:   
[PR-125-locust_tests_1.zip](https://github.com/jetbridge/cdk-nextjs/files/12234876/PR-125-locust_tests_1.zip)

khuezy replied:
For SSG, it makes sense that S3 first would be faster. Now I'm second guessing the actual cause of my big latency by hitting S3 first for SSR.

kevin-mitchell replied:
I suppose that's true. For these tests ("GENERATED" === non-SSG, these were rendered SSR though (I think?)), the difference between S3 first vs Lambda is "only" 23ms. If I'm honest though I don't have enough experience with this setup to know if that's meaningful, or noise.

![Screenshot 2023-08-01 at 9 06 45 PM](https://github.com/jetbridge/cdk-nextjs/assets/21270844/ef5796f5-7928-4253-a90c-28f8415b2c17)

khuezy replied:
For SSG, that would be cached by CDN anyways right? So it would be better to prefer lambda first on SSR. Maybe we can do what Mischa suggested and look up the routes manifest to route accordingly for best performance.

khuezy replied:
If it's only 20ms, then that's not worth doing... but I swear I was getting 300ms the last time I used it. Maybe Nextjs improved since then? Or maybe I did something really stupid before.

kevin-mitchell replied:
@khuezy actually, re:

> For SSG, it makes sense that S3 first would be faster. 

The more I think about this, the more the numbers don't make sense to me. I don't think the SSG files exist in S3 anymore (at least not the first S3 bucket), right? So the first call to S3 I'd think would be a miss, because those SSG files only exist in the `cache` directory, and the S3 bucket setup as an origin in CloudFront are just the static asset files?

khuezy replied:
yea you're right, the .html files are in the cache folder.

revmischa replied:
> If it's only 20ms, then that's not worth doing... but I swear I was getting 300ms the last time I used it. Maybe Nextjs improved since then? Or maybe I did something really stupid before.

It would greatly depend on what region your S3 bucket is in, I imagine.

khuezy replied:
Hmm good point, I'm in west coast and my app was deployed to us-east-1. I think I was running it on the edge too, so that DB round trip might be why I was getting bad latency.

revmischa replied:
My assumption is that CloudFront is in us-east-1, so if your S3 bucket is there then the origin request from CF to S3 should be low latency enough to not greatly impact request time.
If we were to make a Lambda@Edge function that checked for a file in S3, that edge function would possibly have extra latency since it could be run from any AWS POP.

So my point is I would expect using S3 as the primary origin and falling back to lambda if we get a 403/404 to have reasonable performance. Which is why it was originally implemented this way. The 23ms latency to check the S3 origin first sounds pretty reasonable to me. 

Obviously all of these approaches have trade-offs. I think the maximum flexibility and robustness of trying S3 first is worth the extra latency for those requests, and it's always possible to provide some route prefixes that can be routed straight to lambda. Maybe that would be a handy configuration option: `directAppRoutes: string[]` or somesuch, which creates behaviors that skip the fallback origin and go straight to the lambda. Then as a user of this module I could plug in route prefixes that I know are not static files.

bestickley replied:
Great work, @kevin-mitchell!
If it's only 10s of ms difference, I agree with @revmischa:
>  I think the maximum flexibility and robustness of trying S3 first is worth the extra latency for those requests

khuezy replied:
Cool so we're all on the same consensus. If a user experiences longer latency due to edge roundtrip to a distant bucket, they can perhaps override the fallback.

kevin-mitchell replied:
Sounds good! I went ahead and pushed a commit to swap the behavior here. This should resolve the issue with direct requests to static assets (e.g. `<img src="/blah.png" ... />` for `/public/blah.png`) failing.

bestickley commented:
@kevin-mitchell, the "Error related to SST in logs for Server Function" doesn't make sense to me. There should be no code related to SST in this construct. Were you referring to SST's construct?

bestickley commented:
@kevin-mitchell, any other outstanding issues? Do all routes work on the test website?

kevin-mitchell commented:
@bestickley in short, yes, there are still outstanding issues with the test website(s). It's taken me so long to reply because I'm trying to figure out what is actually an issue with this PR vs what is a prior issue / incompatibility between this project and the Vercel app playground / open-next example sites. 

More details below, but I wonder if it might be a good idea to do something like:

1. Open a new PR against main that does nothing but pegs the default open-next version to 1.4 so that somebody could always go back to that tagged release of this project and at least having things working out of the box
2. Either create a "v2-open-next" work branch and then update my PR here to merge into that WIP branch, or merge this PR into main and open new tickets / issues to fix the remaining issues. 

Below are some more details on specific issues I've hit, **but again in some cases I'm not certain it's not just an issue with my local setup and how I'm deploying, etc. So take these with a grain of salt**

-----

I'm just not sure if they are new issues, or old issues. Looking at my PR, honestly there aren't all that many changes so I suspect in the end even if this was landed as is it would be in better shape than before, BUT here are the issues I'm aware of:

## General

* In general it seems that SSG pages when redeployed don't have cache cleared. I'm not sure if this is a new issue, or an existing issue, but (for example) <domain>/ssg.html will be cached at the CloudFront level for a very long time (days?), and so if the site is regenerated and there is a change, this cached item will be held onto. Maybe this is expected, I'm not really sure what the solution is. I don't _think_ any of my changes should have impacted this, but maybe.

## [open-next example](https://github.com/serverless-stack/open-next/tree/main/example)
* Currently anything in `/public` directly referenced does not get served. For reasons I'm not sure about the list of `public-files.json` is not being generated by open-next during the build phase. Part of the mystery as to "why don't we need the S3 asset fallback in the default origin" is because of this: https://github.com/serverless-stack/open-next/blob/274d446ed7e940cfbe7ce05a21108f4c854ee37a/packages/open-next/src/adapters/server-adapter.ts#L93-L101 - basically open-next creates a list of files in the `public` (by default) directory of your app at build time, which is shipped during deployment. But for <reasons I don't yet understand> this isn't being generated currently.

## [Vercel app playground](https://github.com/vercel/app-playground)

* I wonder if this is deployed anywhere with SST. I'd be curious to know how it performs. 
* Some nested layouts aren't working - it seems random, and I wonder if it could have something to do with rate limiting somewhere in my new "free tier" account, or another default somewhere. Either way, it's a bad experience and needs to be better understood and resolved.
* Very frustrating, but ISR seems to work most of the time, but other times it does not. I feel like I'm loosing it a bit because I've watched this work many times, and I know it works with open-next example.

kevin-mitchell commented on src/NextjsRevalidation.ts:
```diff
@@ -0,0 +1,61 @@
+    });
+    const consumer = new Function(this, 'RevalidationFunction', {
+      description: 'Next.js revalidation function',
+      handler: 'index.handler',
```
Sorry I might misunderstand your question, but `.open-next/revalidation-function` is the short answer: https://github.com/serverless-stack/open-next/blob/main/README.md?plain=1#L65

kevin-mitchell commented on src/NextjsRevalidation.ts:
```diff
@@ -0,0 +1,61 @@
+    const server = props.serverFunction.lambdaFunction;
+    server?.addEnvironment('REVALIDATION_QUEUE_URL', queue.queueUrl);
+    server?.addEnvironment('REVALIDATION_QUEUE_REGION', Stack.of(this).region);
+    queue.grantSendMessages(server?.role!);
```
Good call - actually I don't think the `!` is needed in this case, so I'll just go with `queue.grantSendMessages(server);`

kevin-mitchell commented on src/NextjsRevalidation.ts:
```diff
@@ -0,0 +1,61 @@
+  constructor(scope: Construct, id: string, props: RevalidationProps) {
+    super(scope, id);
+
+    if (!props.nextBuild) return;
```
You are correct of course, thanks!

kevin-mitchell commented:
OK so current things that are not working that I'm aware of:

### app playground

1. the "streaming" feature for both edge and node aren't working 
![Screenshot 2023-08-02 at 5 52 07 PM](https://github.com/jetbridge/cdk-nextjs/assets/21270844/2ec393b9-6a15-4bd2-b8a1-b4b2779de41c)

2. "Randomly" I see this issue when navigating through pages quickly
![Screenshot 2023-08-02 at 5 52 07 PM](https://github.com/jetbridge/cdk-nextjs/assets/21270844/28be4559-da11-4100-82f3-f396d822d152)

3. Cache isn't invalidate when deploying, so `/random/cached/page` won't be invalidated during deployment if it changes. Should we add something to invalidate `/*` each deployment?

khuezy commented:
1) Streaming isn't currently supported, Conico has a branch to experiment on https://github.com/serverless-stack/open-next/issues/79.

2) Can you check logs

3) Yes, we should invalidate `/*` on deployment, but I thought cdk-next already did that?

I'll be on tomorrow morning PST.

bestickley commented:
@kevin-mitchell
For 2. above, did you accidentally pasted wrong screenshot? I see the same screenshot as 1.
For 3. above, the code for invalidating is [here](https://github.com/jetbridge/cdk-nextjs/blob/main/assets/lambda/S3EnvRewriter.ts#L109-L125). It looks like we're not invalidating "/*" but rather all files that are referenced in `event.ResourceProperties.s3Keys`. You can add a `console.log(event.ResourceProperties)` [here](https://github.com/jetbridge/cdk-nextjs/blob/main/assets/lambda/S3EnvRewriter.ts#L58C24-L58C48) to check to see if the file that you're seeing that is being incorrectly cached is or isn't in that list. It looks like `s3Keys` is a list of all the static files generated by open-next. See code [here](https://github.com/jetbridge/cdk-nextjs/blob/main/src/NextjsAssetsDeployment.ts#L181-L200).

khuezy commented:
@kevin-mitchell https://github.com/jetbridge/cdk-nextjs/issues/129, this will affect the fallback stuff we discussed.

bestickley commented:
@kevin-mitchell, I'm planning on resolving the remaining issues and then resolving the origin group fallback behavior issue in #129. Server actions are an important feature I think we should support. I'll keep you updated.

kevin-mitchell commented:
@bestickley @khuezy 

RE: Steaming isn't supported - good, I won't worry about having broken it then :) 

RE: Error in logs - I haven't been able to reproduce. I suspect it's a concurrency issue with Lambda / AWS, though I'm not sure why I would have hit the limit.

RE: Invaliding "/*" on deployment - previously, as @bestickley mentioned perhaps this function was handled by invalidating the S3 "assets" bucket contents on deployment. Now this SSG content is inside a different bucket so at least to me it appears these aren't invalidated. I added a new config parameter to allow the full invalidation to be skipped, but defaulted to invalidating `/*` because it will result in the behavior most people would likely expect I think.

bestickley commented:
@kevin-mitchell, on 2nd thought, how about we get this branch merged, then we can address origin group issue in another PR. I want you to "get credit" for this great work. I will test out your changes on this branch tomorrow morning with the open-next test app and if everything looks good I'll merge it.

bestickley commented on src/NextjsBase.ts:
```diff
@@ -72,6 +72,13 @@ export interface NextjsBaseProps {
+   * This can be set to true to skip the full cache invalidation, which
+   * could be important for some users.
+   */
+  readonly skipFullInvalidation?: boolean;
```
So has this never worked? Any files in public/ that were changed but had same name wouldn't get updated in CloudFront cache?

Also, are we sure we want default full invalidation? It makes sense to me, but I want to make sure. @revmischa, can you provide any insight into history of this?

revmischa replied:
It may have not been hooked up by default, as I recall I had some circular dependency issues here.
It's fine to do a full invalidation, I don't think there would be any problem with that. I was just thinking that since the NextJS static file names have hashes in them, it probably isn't strictly necessary to invalidate the cache for static files since the files that change would get new names. Probably better to just invalidate `/*`

kevin-mitchell replied:
I'm _assuming_ that this worked before because the rewriter invalidates the paths in the public S3 bucket. I'm not 100% on this, but I would imagine it would have been called out if it wasn't working before (or perhaps people just were happy creating invalidations in CloudFront manually?).

What I am certain of though is that in my testing at least, without this invalidation the CloudFront cache for SSG pages was not being cleared on deployment.

As for whether the default is appropriate, or whether there needs to be a flag at all... I could be easily convinced either way. On one hand, my honest opinion is somebody using this for a larger, complex project will likely want a more carefully planned and executing cache flushing policy / system, so it's likely some modification of this would happen anyway, so perhaps adding the "skipFullInvalidation" flag is extra noise. On the other hand, it's possible some people would like the option of quickly disabling the flush so they can do it manually and having this config option might be handy.

bestickley replied:
Agreed. Let's keep it.

bestickley commented on src/NextjsLambda.ts:
```diff
@@ -22,6 +22,12 @@ function getEnvironment(props: NextjsLambdaProps): { [name: string]: string } {
+    ...{
+      CACHE_BUCKET_NAME: props.cacheBucket?.bucketName || '',
+      CACHE_BUCKET_REGION: props.cacheBucket ? Stack.of(props.cacheBucket).region : '',
+      // Note we don't need a CACHE_BUCKET_KEY_PREFIX because we're using a separate bucket for cache
```
I'm not against creating another bucket for the cache, but why not reuse static asset bucket with _cache prefix like SST? The smaller footprint for this construct, the better, right?

revmischa replied:
Is there any possibility you could have cached data that you would not want to be publicly accessible? Can people have internal pages cached for example? Would the filenames be guessable?

kevin-mitchell replied:
I originally was actually using `_cache`, [but the `open-next` recommended architecture](https://github.com/serverless-stack/open-next/blob/main/README.md) has them as separate buckets. I sort of like the idea of being able to define separate behavior for these buckets. That said, I did realize SST seems to use the same bucket. I don't have enough experience with the trade-offs here to have a strong opinion, but those were my reasons.

bestickley replied:
@revmischa, I don't think so, because the point of the cache is to make pages ready that are actually being accessed by users. It wouldn't make sense to cache pages that aren't being served by users.

bestickley replied:
@kevin-mitchell, I think the open-next diagram's duplicate bucket is only for simplicity of displaying. I'd prefer to same bucket because I think it semantically makes sense for the cached assets to be in the assets bucket but I'm not hard pressed on this opinion.

kevin-mitchell replied:
Their language [here](https://d328emyhtjj3hd.cloudfront.net) (emphasis mine) to me at least makes it sound like re-using the same bucket is the norm:

> Typically, asset files are uploaded to the root of the bucket. However, you **might** want to store them in a subfolder of the bucket, for instance, when:
> using a pre-existing bucket; or
> storing both assets and cache files in the same bucket.

To me, "cache" === "can be destroyed without breaking things", vs the "assets" directory is genuinely static as far as I can tell - it's more of a "static deployment artifact" bucket. 

(Hopefully this doesn't come off as argumentative. I'm 100% just a guest here and will happily modify this, and really don't have any strong feelings on this. It's possibly I'm partially motivated by the fact that I'll be away from a computer for the next few days and might not get a chance to update and test changing things - the main impact I can see is if we use the sub-directory in the single bucket then the "rewriter" comes into play as it acts on those static files / paths in the assets bucket I believe. Also, I'm not 100% sure how multiple deployments to the same bucket impact each other. None of this is "hard" exactly, just not sure I'll get to it until next week).

bestickley replied:
@kevin-mitchell, no worries at all! I'm working on using the same bucket. I don't think it will impact the rewriter. I'll make sure it works with the deployment construct.

bestickley commented on src/NextjsRevalidation.ts:
```diff
@@ -0,0 +1,64 @@
+}
+
+/**
+ * Builds the system for revaluating Next.js resources. This includes a Lambda function handler and queue system.
```
typo: revalidating

bestickley commented:
@kevin-mitchell, in my testing, I've run across an issue where the `CACHE_*` env vars aren't set if I specify `defaults.lambda.environment`. Could you update the function instantiation in NextjsLambda.ts to below? See comment for details.

```ts
const fn = new Function(scope, 'ServerHandler', {
      memorySize: functionOptions?.memorySize || DEFAULT_LAMBA_MEMORY,
      timeout: functionOptions?.timeout ?? Duration.seconds(10),
      runtime: LAMBDA_RUNTIME,
      handler: path.join('index.handler'),
      code,
      // prevents "Resolution error: Cannot use resource in a cross-environment
      // fashion, the resource's physical name must be explicit set or use
      // PhysicalName.GENERATE_IF_NEEDED."
      functionName: Stack.of(this).region !== 'us-east-1' ? PhysicalName.GENERATE_IF_NEEDED : undefined,
      ...functionOptions,
      // `environment` needs to go after `functionOptions` b/c if
      // `functionOptions.environment` is defined, it will override
      // CACHE_* environment variables which are required
      environment,
    });
```

bestickley commented:
I've verified that ISR works with the Next.js App Router example! All other pages seem to be working too except for loading. @kevin-mitchell, was loading working for you in App Router example app?

kevin-mitchell commented:
> I've verified that ISR works with the Next.js App Router example! All other pages seem to be working too except for loading. @kevin-mitchell, was loading working for you in App Router example app?

![output](https://github.com/jetbridge/cdk-nextjs/assets/21270844/12d257a4-7a10-4533-9951-501db692c6e0)

As far as I know it's working for me - at least based on my understanding. Hopefully this gif either looks "right" or "wrong" - but I just deployed what's out there and this is what I'm seeing.

revmischa commented:
Awesome, if you want to publish a new (major?) version then let's go for it whenever you're ready.

kevin-mitchell commented:
FYI I'm heading out to do some backcountry camping this afternoon. Normally I'd try to make sure I was around in case any fast-follow type fixes were needed, but I'll be "out of pocket" as they say from this afternoon with limited connectivity options.

bestickley commented:
@kevin-mitchell, no worries! Enjoy your trip!
@revmischa, what do you think about doing a pre-release? There are a couple of other updates I'd like to make in the next week that might contain breaking changes so it would be nice to batch them.

bestickley commented:
@kevin-mitchell, I'm glad the loading is working for you! It's not working for me or @khuezy. Not sure why, but not a show stopper IMO.
